### PR TITLE
Avoid zero-sum normalization in GP initial-point selection

### PR DESCRIPTION
**Pull request by tandede**
## Motivation

The mixed acquisition optimizer converts preliminary acquisition values into roulette-selection probabilities after removing the best point. When every remaining exponential weight is zero, either because the values are negative infinity or because they underflow, the probability sum is zero. Dividing by that sum emits a `RuntimeWarning` and produces NaNs even though the existing convergence path can correctly continue with only the best point.

## Description of the changes

- Normalize the roulette probabilities only when their sum is positive.
- Leave an all-zero probability vector unchanged so the existing nonzero-count logic reduces the number of local searches.
- Add a regression test with a degenerate acquisition function whose non-best candidates all have zero probability, with `RuntimeWarning` promoted to an error.

## Validation

- `pytest tests/gp_tests -q` (297 passed)
- `pytest tests/samplers_tests/test_gp.py -q` (14 passed, 1 skipped)
- `ruff format --check optuna/_gp/optim_mixed.py tests/gp_tests/test_optim_mixed.py`
- `ruff check optuna/_gp/optim_mixed.py tests/gp_tests/test_optim_mixed.py`
- `mypy optuna/_gp/optim_mixed.py tests/gp_tests/test_optim_mixed.py`

### CODE REVIEW

**Comment by nabenabe0928** [comment]
Did the issue happen?
If so, please first provide a minimal example to reproduce the issue.
By design, zeros won't Happen mathematically.

We will close this issue unless the issue reproduces.

**Comment by tandede** [comment]
> Did the issue happen? If so, please first provide a minimal example to reproduce the issue. By design, zeros won't Happen mathematically.
> 
> We will close this issue unless the issue reproduces.

Yes, I reproduced this with the actual LogEI implementation. The zero is caused by floating-point underflow, not by the mathematical definition of the weights.

Using the existing `test_after_convergence` setup and four real Sobol candidates, the finite LogEI values are approximately:

`[-5822658.69, -8226.68, -1153076.06, -3099331.28]`

After subtracting the maximum value, `np.exp()` underflows the other three weights to exactly `0.0`. The best candidate is then assigned a zero weight to exclude it from roulette selection, leaving a total probability of `0.0`. The previous normalization consequently emits `RuntimeWarning: invalid value encountered in divide`.

I replaced the synthetic regression test with this real LogEI case. The 296 GP tests and the 14 GPSampler tests pass with the fix.

**Comment by nabenabe0928** [comment]
@tandede 
Could please provide a reproducing code?
Concretely an objective function.

**Comment by tandede** [comment]
> @tandede Could please provide a reproducing code? Concretely an objective function.

Yes. Here is a reproducer using `study.optimize()` and a concrete objective:

```python
import warnings

import optuna


xs = [
    *(i / 10 for i in range(1, 11)),
    *(i / 1e5 for i in range(1, 21)),
    *([0.0] * 10),
]

sampler = optuna.samplers.GPSampler(seed=0, n_startup_trials=0)

# Keep the Sobol candidate set small so the numerical case is
# deterministic and quick to reproduce. The GP and LogEI are unchanged.
sampler._n_preliminary_samples = 4
sampler._n_local_search = 2

study = optuna.create_study(direction="minimize", sampler=sampler)

for x in xs:
    study.enqueue_trial({"a": x})


def objective(trial: optuna.Trial) -> float:
    return trial.suggest_float("a", 0.0, 1.0)


with warnings.catch_warnings():
    warnings.simplefilter("error", RuntimeWarning)
    study.optimize(objective, n_trials=len(xs) + 1)
```

On `4badf9a1b`, trial 40 fails with:

```text
RuntimeWarning: invalid value encountered in divide
```

The traceback points to `probs /= probs.sum()` in `optimize_acqf_mixed()`.

On this PR’s current head, all 41 trials complete successfully. The two private sampler assignments only make the underflow deterministic and fast to reproduce; the example still uses the actual `GPSampler`, fitted GP, LogEI acquisition function, Sobol candidates, and objective path.

**Comment by nabenabe0928** [comment]
```
# Keep the Sobol candidate set small so the numerical case is
# deterministic and quick to reproduce. The GP and LogEI are unchanged.
sampler._n_preliminary_samples = 4
sampler._n_local_search = 2
```

Please reproduce without them.
Especially, both of them don't make sense practically.

@tandede

**Comment by tandede** [comment]
> ```
> # Keep the Sobol candidate set small so the numerical case is
> # deterministic and quick to reproduce. The GP and LogEI are unchanged.
> sampler._n_preliminary_samples = 4
> sampler._n_local_search = 2
> ```
> 
> Please reproduce without them. Especially, both of them don't make sense practically.
> 
> @tandede

Thanks, you are right that overriding those two private settings does not demonstrate the behavior under the practical default path. I removed both overrides and reproduced it with the default `GPSampler` configuration (`2048` preliminary Sobol samples and `10` local searches).

```python
import warnings

import optuna


sampler = optuna.samplers.GPSampler(seed=22, n_startup_trials=0)
study = optuna.create_study(direction="minimize", sampler=sampler)

xs = [
    *(i / 10 for i in range(1, 11)),
    *(i / 1e5 for i in range(1, 21)),
    *([0.0] * 60),
]

for x in xs:
    study.enqueue_trial({"a": x})


def objective(trial: optuna.Trial) -> float:
    return trial.suggest_float("a", 0.0, 1.0)


with warnings.catch_warnings():
    warnings.simplefilter("error", RuntimeWarning)
    study.optimize(objective, n_trials=len(xs) + 1)
```

On the pre-fix commit `4badf9a1b`, trial 90 (the 91st trial) fails at:

```text
probs /= probs.sum()
RuntimeWarning: invalid value encountered in divide
```

With the default 2048 Sobol candidates, the largest and second-largest finite LogEI values are approximately `-108.30` and `-893.55`. Their difference is about `785.26`, so every non-best exponential weight underflows to `0.0`. After the best candidate is excluded, the probability sum is exactly zero.

On the current PR head, all 91 trials complete. I also updated the regression test to use the default optimizer arguments and pushed it in `f0b24dad6`. The 296 GP tests pass, and the GPSampler test file reports 14 passed and 1 skipped.

File: optuna/_gp/optim_mixed.py [hunk 1]
```diff
@@ -308,7 +308,9 @@ def optimize_acqf_mixed(
     # We use a modified roulette wheel selection to pick the initial param for each local search.
     probs = np.exp(f_vals - f_vals[max_i])
     probs[max_i] = 0.0  # We already picked the best param, so remove it from roulette.
-    probs /= probs.sum()
+    probs_sum = probs.sum()
+    if probs_sum > 0.0:
+        probs /= probs_sum
```
Comment (nabenabe0928) [inline optuna/_gp/optim_mixed.py]:
```suggestion
    probs = probs / probs_sum if (probs.sum() := probs_sum) > 0 else probs
```

File: tests/samplers_tests/test_gp.py [hunk 2]
```diff
@@ -41,9 +41,19 @@ def test_after_convergence(caplog: LogCaptureFixture) -> None:
...
+    probs[max_i] = 0.0
+    assert np.isfinite(f_vals).all()
+    assert not np.any(probs)
+
     caplog.clear()
     optuna.logging.enable_propagation()
-    optim_mixed.optimize_acqf_mixed(acqf_params, rng=np.random.RandomState(42))
+    with warnings.catch_warnings():
+        warnings.simplefilter("error", RuntimeWarning)
+        optim_mixed.optimize_acqf_mixed(acqf_params, rng=np.random.RandomState(22))
```
Comment (nabenabe0928) [inline tests/samplers_tests/test_gp.py]: Please revert it.
This change is not necessary.

File: tests/samplers_tests/test_gp.py [hunk 3]
```diff
@@ -41,9 +41,19 @@ def test_after_convergence(caplog: LogCaptureFixture) -> None:
     acqf_params = acqf_module.LogEI(
         gpr=gpr, search_space=search_space, threshold=np.max(score_vals)
     )
+    sampled_xs = search_space.sample_normalized_params(2048, rng=np.random.RandomState(22))
+    f_vals = acqf_params.eval_acqf_no_grad(sampled_xs)
+    max_i = np.argmax(f_vals)
+    probs = np.exp(f_vals - f_vals[max_i])
+    probs[max_i] = 0.0
+    assert np.isfinite(f_vals).all()
+    assert not np.any(probs)
```
Comment (nabenabe0928) [inline tests/samplers_tests/test_gp.py:77]: Could you please isolate this test?

File: optuna/_gp/optim_mixed.py (hunk 1 above)
Comment (tandede) [reply optuna/_gp/optim_mixed.py]: Thanks. I applied the intended conditional expression using (probs_sum := probs.sum()), which keeps the assignment target valid while preserving the suggested one-line form.

File: tests/samplers_tests/test_gp.py (hunk 2 above)
Comment (tandede) [reply tests/samplers_tests/test_gp.py]: Reverted. The warning filter has been removed, and the existing test_after_convergence test has been restored to its original behavior.

File: tests/samplers_tests/test_gp.py (hunk 3 above)
Comment (tandede) [reply tests/samplers_tests/test_gp.py:77]: Done. I moved the LogEI underflow case into a standalone test_after_convergence_with_zero_sum_probabilities test. The existing convergence test is unchanged, and the new regression still uses the default 2048 preliminary samples and 10 local searches.

**Comment by tandede** [comment]
Hi @nabenabe0928, all requested changes have been addressed:

- the issue is reproduced with the default GPSampler settings;
- the zero-sum LogEI regression is isolated in its own test;
- the unnecessary change to the existing convergence test was reverted;
- all 13 reported checks are passing.

When you have time, could you please take another look? Thank you.

**Comment by not522** [comment]
@nabenabe0928 Could you continue reviewing this?
@kAIto47802 Could you review this PR?
